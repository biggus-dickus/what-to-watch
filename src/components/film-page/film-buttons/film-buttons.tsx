import * as React from 'react';
import {Link} from 'react-router-dom';

import RouteConfig from '../../../config/routes';

interface Props {
  filmId: number,
  isAdded: boolean,
  onAddToWatchList: (id: number) => Promise<any>,
  onRemoveFromWatchList: (id: number) => Promise<any>
}

export const FilmButtons = ({filmId, isAdded, onAddToWatchList, onRemoveFromWatchList}: Props): React.ReactElement => {
  const [isFavourite, updateIsFavourite] = React.useState(isAdded);
  React.useEffect(() => updateIsFavourite(isAdded), [isAdded]);

  let btnIcon = (
    <svg viewBox="0 0 19 20" width="19" height="20" data-test="at-is-not-added">
      <use xlinkHref="img/sprite/sprite.svg#add" />
    </svg>
  );

  let btnText = `My list`;
  let clickHandler = (id) => onAddToWatchList(id)
    .then((resp) => {
      if (resp.is_favorite) {
        updateIsFavourite(resp.is_favorite);
      }
    });

  if (isFavourite) {
    btnIcon = (
      <svg viewBox="0 0 18 14" width="18" height="14" data-test="at-is-added">
        <use xlinkHref="img/sprite/sprite.svg#in-list" />
      </svg>
    );

    btnText = `In my list`;
    clickHandler = (id) => onRemoveFromWatchList(id)
      .then(() => updateIsFavourite(false));
  }

  return (
    <div className="movie-card__buttons">
      <button className="btn btn--play movie-card__button" type="button">
        <svg viewBox="0 0 19 19" width="19" height="19">
          <use xlinkHref="img/sprite/sprite.svg#play-s" />
        </svg>
        <span>Play</span>
      </button>

      <button
        className="btn btn--list movie-card__button"
        data-test="at-my-list-btn"
        type="button"
        onClick={clickHandler.bind(null, filmId)}>
        {btnIcon}
        <span>{btnText}</span>
      </button>

      <Link
        to={RouteConfig.ADD_REVIEW.replace(`:id`, `` + filmId)}
        className="btn movie-card__button">
        Add review
      </Link>
    </div>
  );
};

export default FilmButtons;
