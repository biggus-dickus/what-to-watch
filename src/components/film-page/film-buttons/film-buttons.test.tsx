import * as React from 'react';
import {StaticRouter as Router} from 'react-router';
import * as renderer from 'react-test-renderer';
import * as Adapter from 'enzyme-adapter-react-16';
import {configure, shallow} from 'enzyme';

import {FilmButtons} from './film-buttons';

configure({adapter: new Adapter()});


const props = {
  filmId: 1,
  isAdded: false,
  onWatchListToggle: jest.fn(),
  onPlayerLaunch: jest.fn()
};

describe(`FilmButtons test suite`, () => {
  it(`<FilmButtons /> should render correctly`, () => {
    expect(renderer.create(<Router><FilmButtons {...props} /></Router>).toJSON()).toMatchSnapshot();
  });

  it(`"My list" button should display a special icon and text when the currently rendered film IS NOT in user's favourites`, () => {
    const wrapper = shallow(<FilmButtons {...props} />);

    expect(wrapper.find(`[data-test="at-is-not-added"]`)).toHaveLength(1);
    expect(wrapper.find(`[data-test="at-my-list-btn"]`).text()).toEqual(`My list`);
  });

  it(`"My list" button should display a special icon and text when the currently rendered film IS in user's favourites`, () => {
    const newProps = {...props, isAdded: true};
    const wrapper = shallow(<FilmButtons {...newProps} />);

    expect(wrapper.find(`[data-test="at-is-added"]`)).toHaveLength(1);
    expect(wrapper.find(`[data-test="at-my-list-btn"]`).text()).toEqual(`In my list`);
  });
});
