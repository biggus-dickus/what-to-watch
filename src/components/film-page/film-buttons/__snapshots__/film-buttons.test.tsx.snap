// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilmButtons test suite <FilmButtons /> should render correctly 1`] = `
<div
  className="movie-card__buttons"
>
  <button
    className="btn btn--play movie-card__button"
    data-test="at-play-btn"
    onClick={[MockFunction]}
    type="button"
  >
    <svg
      height="19"
      viewBox="0 0 19 19"
      width="19"
    >
      <use
        xlinkHref="img/sprite/sprite.svg#play-s"
      />
    </svg>
    <span>
      Play
    </span>
  </button>
  <button
    className="btn btn--list movie-card__button"
    data-test="at-my-list-btn"
    onClick={[Function]}
    type="button"
  >
    <svg
      data-test="at-is-not-added"
      height="20"
      viewBox="0 0 19 20"
      width="19"
    >
      <use
        xlinkHref="img/sprite/sprite.svg#add"
      />
    </svg>
    <span>
      My list
    </span>
  </button>
  <a
    className="btn movie-card__button"
    href="/film/1/review"
    onClick={[Function]}
  >
    Add review
  </a>
</div>
`;
